src/main/java/net/minecraftforge/registries/IForgeRegistry.java

```diff
@@ -34,17 +39,24 @@
     boolean containsKey(ResourceLocation key);
     boolean containsValue(V value);
     boolean isEmpty();
+    boolean supportsTags();
 
     @Nullable V getValue(ResourceLocation key);
     @Nullable ResourceLocation getKey(V value);
     @Nullable ResourceLocation getDefaultKey();
-    @Nullable Optional<ResourceKey<V>> getResourceKey(V value);
+    @NotNull Optional<ResourceKey<V>> getResourceKey(V value);
 
-    @Nonnull Set<ResourceLocation>         getKeys();
-    @Nonnull Collection<V>                 getValues();
-    @Nonnull Set<Entry<ResourceKey<V>, V>> getEntries();
+    @NotNull Set<ResourceLocation>         getKeys();
+    @NotNull Collection<V>                 getValues();
+    @NotNull Set<Entry<ResourceKey<V>, V>> getEntries();
 
-    @Nonnull Codec<V> getCodec();
+    @NotNull Codec<V> getCodec();
+
+    @NotNull Optional<Holder<V>> getHolder(ResourceKey<V> key);
+    @NotNull Holder<V> getHolderOrThrow(ResourceKey<V> key);
+    @NotNull HolderSet.Named<V> getOrCreateTag(TagKey<V> name);
+    @NotNull Optional<HolderSet.Named<V>> getTag(TagKey<V> name);
+    @NotNull Iterable<Holder<V>> getTagOrEmpty(TagKey<V> name);
```

Comment:
I believe it may also make sense to expose the following methods that also exist on `MappedRegistry`:
- `boolean isKnownTagName(TagKey)`
- `Stream<Pair<TagKey<T>, HolderSet.Named<T>>> getTags()`
- `Stream<TagKey<T>> getTagNames()`